File: backend/app/models/course.py
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, Text
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.core.database import Base


class Course(Base):
    __tablename__ = "course"

    id = Column(Integer, primary_key=True, autoincrement=True)
    user_id = Column(Integer, ForeignKey("user.id"), nullable=False)
    name = Column(String(128), nullable=False)
    code = Column(String(64), nullable=True)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())

    experiments = relationship("Experiment", back_populates="course", cascade="all, delete-orphan")
    user = relationship("User", back_populates="courses")


class Clazz(Base):
    __tablename__ = "clazz"

    id = Column(Integer, primary_key=True, autoincrement=True)
    user_id = Column(Integer, ForeignKey("user.id"), nullable=False)
    name = Column(String(128), nullable=False)
    grade = Column(String(32), nullable=True)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())

    reports = relationship("Report", back_populates="clazz")


class Experiment(Base):
    __tablename__ = "experiment"

    id = Column(Integer, primary_key=True, autoincrement=True)
    user_id = Column(Integer, ForeignKey("user.id"), nullable=False)
    course_id = Column(Integer, ForeignKey("course.id", ondelete="CASCADE"), nullable=False)
    title = Column(String(256), nullable=False)
    description = Column(Text, nullable=True)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())

    course = relationship("Course", back_populates="experiments")
    reports = relationship("Report", back_populates="experiment", cascade="all, delete-orphan")
    check_tasks = relationship("CheckTask", back_populates="experiment", cascade="all, delete-orphan")
    user = relationship("User", back_populates="experiments")

File: frontend/src/api/course.js
import { get, post, del } from './http'

export function getCourses() {
  return get('/course/courses')
}

export function createCourse(name, code = '') {
  return post('/course/courses', { name, code })
}

export function deleteCourse(id) {
  return del(`/course/courses/${id}`)
}

export function getClasses() {
  return get('/course/classes')
}

export function createClass(name, grade = '') {
  return post('/course/classes', { name, grade })
}

export function deleteClass(id) {
  return del(`/course/classes/${id}`)
}

export function getExperiments() {
  return get('/course/experiments')
}

export function createExperiment(courseId, title, description = '') {
  return post('/course/experiments', { courseId, title, description })
}

export function deleteExperiment(id) {
  return del(`/course/experiments/${id}`)
}
